### PR TITLE
Refactor minibatch ADVI and update docstring

### PR DESCRIPTION
**Pull request by taku-y**
I refactored minibatch ADVI and update docstring. The equation of ELBO of minibatch (and autoencoding-VB) ADVI in PyMC3 is written and the input arguments is documented in detail. 

I have tested the function on the example notebooks where `advi_minibatch` is used.

### CODE REVIEW

File: pymc3/variational/advi_minibatch.py [hunk 1]
```diff
@@ -203,52 +216,167 @@ def logp_(z_g):
...
+        \sum_{k=1}^{V_{l}}c_{l}^{k}KL\left[
+            q(\mathbf{z}_{i}^{k})||
+            p(\mathbf{z}_{i}^{k}|{\\rm pa}(\mathbf{z}_{i}^{k}))\\right],
+
+    where :math:`{\\rm pa}(v)` denotes the set of parent variables of :math:`v`
+    in the directed acyclic graph of the model.
+
+    When using mini-batches, :math:`c_{o}^{k}` and :math:`c_{l}^{k}` should be
+    set to :math:`N/M`, where :math:`M` is the number of observations in each
+    mini-batch. Other weighting scheme were proposed in (Blundell et al., 2015)
```
Comment (twiecki) [inline pymc3/variational/advi_minibatch.py]: Typo: Another weighting scheme was proposed in...

File: pymc3/variational/advi_minibatch.py [hunk 2]
```diff
@@ -203,52 +216,167 @@ def logp_(z_g):
...
+    parameters of the global variables, :math:`\gamma`, because these are
+    automatically created and updated in this function.
+
+    The following is a list of example notebooks using advi_minibatch:
+
+    - docs/source/notebooks/GLM-hierarchical-advi-minibatch.ipynb
+    - docs/source/notebooks/bayesian_neural_network_advi.ipynb
+    - docs/source/notebooks/convolutional_vae_keras_advi.ipynb
+    - docs/source/notebooks/gaussian-mixture-model-advi.ipynb
+    - docs/source/notebooks/lda-advi-aevb.ipynb
```
Comment (twiecki) [inline pymc3/variational/advi_minibatch.py:381]: This is one amazing doc-string.

**Comment by twiecki** [comment]
@taku-y Refactoring looks good. Is the re-weighting helping with convergence?

I re-pointed this PR to the 3.1 branch and it seems you have to rebase.

In terms of @ferrine's work it seems like the AEVB is not implemented there so it makes sense to move forward here and perhaps merge the larger refactoring (if we think it's an improvement) later. Would still be great if @ferrine could help review this.

**Comment by ferrine** [comment]
I'll probably refactor my code. But first I need to understand the paper about AEVB, it looks interesting.

**Comment by twiecki** [comment]
Sounds good. Did you have a look at this PR?

**Comment by ferrine** [comment]
Just a glance, did not have enough time for a closer look

**Comment by springcoil** [comment]
Had a look at this, I like some of the changes made! Excellent work @taku-y. I can't wait to try this out. I think @ferrine would be a good person to give his view though before a merge.

**Review by ferrine — COMMENTED** [review]
There are minor comments and suggestions for refactoring

File: pymc3/variational/advi_minibatch.py [hunk 3]
```diff
@@ -334,30 +477,22 @@ def get_transformed(v):
...
     replaces = dict()
 
     # Variational parameters for global RVs
     if 0 < len(global_RVs):
-        uw_global_shared, bij = _init_uw_global_shared(start, global_RVs,
-                                                       global_order)
+        uw_global_shared, bij = _init_uw_global_shared(start, global_RVs)
         replaces.update({uw_g: uw_global_shared})
 
     # Variational parameters for local RVs, encoded from samples in
```
Comment (ferrine) [inline pymc3/variational/advi_minibatch.py:503]: BTW this fails if n < 10. That number can be used in debugging 
```python
...
     for i in progress:
         e = f(*next(minibatches))
         elbos[i] = e
         if i % (n // 10) == 0 and i > 0:
...
```
PS the lowest line that was available for coment

File: pymc3/variational/advi_minibatch.py [hunk 4]
```diff
@@ -334,30 +477,22 @@ def get_transformed(v):
...
-    global_RVs = list(set(vars) - set(list(local_RVs) + list(observed_RVs)))
-
-    # Ordering for concatenation of random variables
-    global_order = pm.ArrayOrdering([v for v in global_RVs])
-    local_order = pm.ArrayOrdering([v for v in local_RVs])
+    grvs = list(set(vars) - set(list(local_RVs) + list(observed_RVs)))
+    if global_RVs is None:
+        global_RVs = OrderedDict({v: 1 for v in grvs})
+    else:
+        assert(len(grvs) == len(global_RVs))
```
Comment (ferrine) [inline pymc3/variational/advi_minibatch.py]: Is there is no error message, assertion fail will look really strange

File: pymc3/variational/advi_minibatch.py [hunk 5]
```diff
@@ -334,30 +477,22 @@ def get_transformed(v):
     )
```
Comment (ferrine) [inline pymc3/variational/advi_minibatch.py:478]: `v in ds` does not mean it is transformed, I can provide a custom expression to Deterministic function or they cannot be in local vars? This is not clear for me. Maybe just check `hasattr(var, 'transformed')`?
```python
ds = model.deterministics
 
     def get_transformed(v):
         if v in ds:
             return v.transformed
         return v
```

File: pymc3/variational/advi_minibatch.py [hunk 6]
```diff
@@ -305,6 +437,17 @@ def advi_minibatch(vars=None, start=None, model=None, n=5000, n_mcsamples=1,
...
+    - Kingma, D. P., & Welling, M. (2014).
+      Auto-Encoding Variational Bayes. stat, 1050, 1.
+    - Kucukelbir, A., Ranganath, R., Gelman, A., & Blei, D. (2015).
+      Automatic variational inference in Stan. In Advances in neural
+      information processing systems (pp. 568-576).
+    - Blundell, C., Cornebise, J., Kavukcuoglu, K., & Wierstra, D. (2015).
+      Weight Uncertainty in Neural Network. In Proceedings of the 32nd
+      International Conference on Machine Learning (ICML-15) (pp. 1613-1622).
     """
     theano.config.compute_test_value = 'ignore'
```
Comment (ferrine) [inline pymc3/variational/advi_minibatch.py]: this is a side effect, it should be wrapped in context manager

```python
@with_no_test_val
def advi_minibatch(...):
    ...
```

File: pymc3/variational/advi_minibatch.py (hunk 2 above)
Comment (ferrine) [reply pymc3/variational/advi_minibatch.py:381]: really great

File: pymc3/variational/advi_minibatch.py [hunk 7]
```diff
@@ -203,53 +216,167 @@ def logp_(z_g):
...
+    return elbo, uw_l, uw_g
 
-    - observed_RVs=OrderedDict([(rv, total_size / minibatch_size)])
-    - minibatch_RVs=[rv], total_size=total_size
 
-    where minibatch_size is minibatch_tensors[0].shape[0].
+def advi_minibatch(vars=None, start=None, model=None, n=5000, n_mcsamples=1,
+                   minibatch_RVs=None, minibatch_tensors=None,
+                   minibatches=None, global_RVs=None, local_RVs=None,
+                   observed_RVs=None, encoder_params=[], total_size=None,
```
Comment (ferrine) [inline pymc3/variational/advi_minibatch.py]: mutable default value seems to be modified [here](https://github.com/pymc-devs/pymc3/pull/1573/files#diff-1e408f9f0670474e10d1d854282138d0R521)
You can use
```python
def advi_minibatch(
    ...
    observed_RVs=None, encoder_params=None, total_size=None
    ...
    )
    if encoder_params is None:
        encoder_params = []
```

**Comment by taku-y** [comment]
@twiecki The reweighting of ELBO doesn't help convergence without scheduling of the coefficients as shown in the example of @ferrine. I think the new docstring with the detailed description of API is beneficial for users. In 3.1, it would be better to refactor ADVI with VariationalReplacement by @ferrine.

**Comment by twiecki** [comment]
@taku-y Sounds good. Will wait for you to respond/fix the minor issues @ferrine pointed out.

File: pymc3/variational/advi_minibatch.py (hunk 4 above)
Comment (taku-y) [reply pymc3/variational/advi_minibatch.py]: Fair enough. I will fix it to raise ValueError.

File: pymc3/variational/advi_minibatch.py (hunk 6 above)
Comment (taku-y) [reply pymc3/variational/advi_minibatch.py]: I have not seen this decolator. Do I need to define it or can I import from another module?

File: pymc3/variational/advi_minibatch.py (hunk 7 above)
Comment (taku-y) [reply pymc3/variational/advi_minibatch.py]: I will fix it.

File: pymc3/variational/advi_minibatch.py (hunk 3 above)
Comment (taku-y) [reply pymc3/variational/advi_minibatch.py:503]: I will fix the code so that it accepts n < 10.

Comment (taku-y) [reply pymc3/variational/advi_minibatch.py:503]: And add a test for n < 10.

File: pymc3/variational/advi_minibatch.py (hunk 5 above)
Comment (taku-y) [reply pymc3/variational/advi_minibatch.py:478]: Right. Using hasattr would be better. I will fix it and run it on the LDA-AEVB notebook for test.

**Comment by taku-y** [comment]
@springcoil @ferrine Thanks for review. I fixed the code according to comments except for the one for transform checking. I will fix it and do test tomorrow.

File: pymc3/variational/advi_minibatch.py (hunk 6 above)
Comment (ferrine) [reply pymc3/variational/advi_minibatch.py]: you can use [this](https://github.com/Theano/Theano/blob/master/theano/configparser.py#L94) function from theano

**Comment by taku-y** [comment]
I fixed checking transformed RVs according to the comment and tested on the notebooks using advi.

**Comment by ferrine** [comment]
@taku-y I tired to do some generalisations concerning variational methods, it looks like [this](https://github.com/ferrine/pymc3/blob/refactor_advi/pymc3/variational/replacements.py) now. SVGD, ADVI and ENH Normalizing flows possibly can be implemented in that unified way

**Comment by twiecki** [comment]
@taku-y Is there anything you still want to change here or is it ready to merge from your end?

**Comment by taku-y** [comment]
@ferrine Great. I want to refactor the code with your functions. It will need solid blocks of time, so I will do it after this PR finishes.

@twiecki I forgot to fix the code where setting testval. I will change it.

**Comment by springcoil** [comment]
This paper might be of interest to you @taku-y https://papers.nips.cc/paper/6581-improving-variational-autoencoders-with-inverse-autoregressive-flow.pdf

**Comment by springcoil** [comment]
Just so I understand @taku-y this is on hold until the code on test val is added in?

**Comment by taku-y** [comment]
@springcoil Thanks for telling me about the paper. It looks improved from the previous version. 

And as you said, now I'm modifying the code on test val.

**Comment by twiecki** [comment]
@taku-y Is this ready to merge?

**Comment by taku-y** [comment]
@twiecki I think so.

**Comment by twiecki** [comment]
Thanks @taku-y!

**Pull request merged**